### PR TITLE
Extract RowBuffer class for SparseRowMatrix.

### PR DESCRIPTION
**Pull request by reyoung**
* The original SparseRowMatrix use two fields to store each row,
  which make code very confusing. Try to extract a RowBuffer class,
  for SparseRowMatrix data storage, and manage auto-growth logic.

### CODE REVIEW

File: paddle/math/RowBuffer.h [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  RowBuffer(const CpuMemHandlePtr& mem, size_t width)
+      : preallocatedBuf_(mem), width_(width) {}
+
+  /**
+   * @brief resize resize the buffer with rowCount
+   * @param rowCnt number of row. matrix height.
+   */
+  inline void resize(int rowCnt) {
+    if (preallocatedBuf_) {
+      CHECK(preallocatedBuf_->getSize() < rowCnt * width_ * sizeof(real));
```
Comment (hedaoyuan) [inline paddle/math/RowBuffer.h]: 这里写反了吧？

File: paddle/math/RowBuffer.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+  /**
+   * @brief return the width of matrix. a.k.a length of row.
+   * @return width of matrix
+   */
+  inline size_t getWidth() const { return width_; }
+
+private:
+  CpuMemHandlePtr preallocatedBuf_;
+  std::vector<real, AlignedAllocator<real, 32>> rowStore_;
```
Comment (hedaoyuan) [inline paddle/math/RowBuffer.h]: rowStore_是不是直接用CpuMemoryHandle就可以了，只是resize greater的时候需要一次copy。

File: paddle/math/RowBuffer.h [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+   */
+  inline void clear() { rowStore_.clear(); }
+
+  /**
+   * @brief get current number of rows.
+   * @return number of rows.
+   */
+  inline size_t getRowCount() const {
+    if (preallocatedBuf_) {
+      return preallocatedBuf_->getSize() / sizeof(float) / width_;
```
Comment (hedaoyuan) [inline paddle/math/RowBuffer.h]: sizeof(real)

File: paddle/math/RowBuffer.h [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    } else {
+      return rowStore_.size() / width_;
+    }
+  }
+
+  /**
+   * @brief get is this buffer can automatically grow or not.
+   * @return ture if can automacitally grow.
+   */
+  inline bool isAutoGrowth() const { return preallocatedBuf_ == nullptr; }
```
Comment (hedaoyuan) [inline paddle/math/RowBuffer.h]: return !preallocatedBuf_;

File: paddle/math/RowBuffer.h (hunk 2 above)
Comment (reyoung) [inline paddle/math/RowBuffer.h]: 有道理，不过再下一个PR里面加吧。

加了一个TODO注释

File: paddle/math/RowBuffer.h (hunk 4 above)
Comment (reyoung) [inline paddle/math/RowBuffer.h]: done.

File: paddle/math/RowBuffer.h (hunk 3 above)
Comment (reyoung) [inline paddle/math/RowBuffer.h]: done.

File: paddle/math/RowBuffer.h (hunk 1 above)
Comment (reyoung) [inline paddle/math/RowBuffer.h]: done.

**Pull request merged**